### PR TITLE
Adding support for task specific Chroot

### PR DESCRIPTION
**Pull request by devendram**
This PR adds support for task specific chroot. Validation is added for task chroot, agent chroot, and default chroot.  task chroot cannot go outside agent chroot, and agent chroot cannot go outside default chroot.

default chroot > agent chroot > task chroot

### CODE REVIEW

**Comment by devendram** [comment]
valid job spec chroot
task "example" {
    driver = "exec"
    config {
        command = "/bin/sleep"
        args = ["1"]
    }
    chroot_env {
        "/bin/ls" = "/bin/ls"
         "/lib/libiptc.so.0" = "/lib/libiptc.so.0"
    }
}

**Comment by devendram** [comment]
https://github.com/hashicorp/nomad/issues/1822

**Comment by dadgar** [comment]
@devendram Sorry for the delay. There is going to be a refactor in how we build chroots in 0.5.2. I want this to hit master and then we can talk about how to rebase this such that it can be incorporated.

Thanks for the patience

File: client/driver/executor/executor.go [hunk 1]
```diff
@@ -901,3 +902,24 @@ func (e *UniversalExecutor) Signal(s os.Signal) error {
 
 	return nil
 }
+
+func GetChrootEnvTree(chrootEnv map[string]string) *iradix.Tree {
```
Comment (schmichael) [inline client/driver/executor/executor.go]: On master as of #2132 all chroot building code has moved out of the drivers+executor and into TaskRunner.

**Comment by schmichael** [comment]
Thanks for the PR; the code looks great. We're going to focus on the layered filesystem solution to solve speed/space issues with chroots, so I'm going to close this.

I left a couple review comments in case you want to maintain this patch internally until we can land that feature. (And also because it was just a pleasure to review! Great use of radix trees.)

**Comment by jshaw86** [comment]
@schmichael  Per a discussion I had with @dadgar I was concerned about the separation in responsibility and the workflow of these layered chroots. Was hoping to get a little more clarity on how one would go about building out these chroots specific for interpreted environments like node and python.

**Comment by schmichael** [comment]
@jshaw86 We don't have specific plans yet, but there would be a couple layers of the implementation you could use for these sort of situations:

* The first layer being the host filesystem which gets included in the chroot. With zfs/aufs/etc CoW chroots you could install all dependencies globally and not pay any extra space/time cost per-task.
* Obviously `installing all dependencies everywhere` isn't ideal, so the second layer would be to use artifacts to pull in per-task dependencies. You could do that today, but we currently don't support the most common way to ship system libraries your Python/Ruby/Node code relies on: distro packages!

So if for example you had multiple tasks depending on `libpq` being installed you could either include it in your base system image or you could specify it on a per-task basis with artifacts.

Would supporting installing deb/rpms inside a task's chroot through artifacts meet your use case?

**Comment by jshaw86** [comment]
So our requirements in order are:
1. Fast startup time
2. Clear separation between developer and operational responsibility
3. Simple developer interactions with packaging

By fast startup time I mean in the ms range. When we were running vanilla chroot settings we were seeing seconds to get a new chroot. This is why we implemented the agent chroot mapping to at least pair it down per agent.

Currently our ops team works in terraform and packer. Our developers work with nomad spec files and control dist packages which basically package virtual environments whether its python or node. The end goal here is an ops engineer should be able to bring up new hardware without concern of messing up applications and a developer should be able to develop an application that executes within a nomad client without having to be concerned about the hardware its running.

> The first layer being the host filesystem which gets included in the chroot. With zfs/aufs/etc CoW chroots you could install all dependencies globally and not pay any extra space/time cost per-task.

Can we get a subset of the host file system with this strategy, we will certainly never want the entire host chroot but will have a common set of things from the host for all tasks? E.g. what if I want /etc/resolv.conf in my chroot but not /usr/share. Here is a sample nomad client chroot spec:

> Obviously installing all dependencies everywhere isn't ideal, so the second layer would be to use artifacts to pull in per-task dependencies. You could do that today, but we currently don't support the most common way to ship system libraries your Python/Ruby/Node code relies on: distro packages!

So I think this comes down to execution. It feels very heavy to me, when we're basically just trying to take a subset of the host filesystem for a specific tasks. If I'm running a python application I only want python binaries. Same if I'm running node or ruby. In a single line "the task should have exactly what it needs no more no less" and do that in the simplest implementation way with the fastest startup time.

To me having terraform/packer do the heavy lifting of getting binaries onto the host file system and then having nomad spec files just slice what it needs seems really straight forward and provides a nice separation between ops and dev responsibilities. If this is better that's great but I would have to see the workflow end to end I give it my thumbs up.

This is our current agent chroot_env to give you an idea how small we need.
```
  chroot_env {
    \"/lib/x86_64-linux-gnu\" = \"/lib/x86_64-linux-gnu\"
    \"/lib64\" = \"/lib64\"
    \"/usr/lib\" = \"/usr/lib\"
    \"/etc/ld.so.cache\" = \"/etc/ld.so.cache\"
    \"/etc/ld.so.conf\" = \"/etc/ld.so.conf\"
    \"/etc/ld.so.conf.d\" = \"/etc/ld.so.conf.d\"
    \"/etc/resolvconf\" = \"/etc/resolvconf\"
    \"/run/resolvconf/resolv.conf\" = \"/etc/resolv.conf\"
    \"/usr/bin/nodejs\" = \"/usr/bin/node\"
    \"/usr/bin/curl\" = \"/usr/bin/curl\"
    \"/etc/ssl\" = \"/etc/ssl\"
    \"/etc/ca-certificates\" = \"/etc/ca-certificates\"
  }
}
```

**Comment by schmichael** [comment]
While I think using an overlay fs to create the chroot would meet your performance requirements I can see how it doesn't meet your other two requirements.

Another feature we'd like to add is a `rootfs` artifact config option to the exec/java drivers to allow downloading an artifact that's an entire root filesystem. Nothing would be copied from the host. The rootfs artifact would be fetched, extracted, task directories like `alloc/` would be created, other artifacts would be fetched, and then your task would run.

This would be similar to a Docker image in that it would contain only what the developers wanted to include. It could even be produced by using `docker export` to create a rootfs tarball from an image.

To further improve task startup times and space requirements the root fs would be cached locally (like Docker images) and an overlay fs used so the underlying rootfs is shared and immutable.

**Comment by schmichael** [comment]
(I should also mention we on the Nomad team really appreciate your input. We know the exec driver has a major deficiency in this area and your feedback is helpful in determining and validating our plans!)

**Comment by jshaw86** [comment]
@schmichael yea @dadgar mentioned the rootfs to me a while back. So the trade offs are the follow as I see them over the job spec.

Advantages
1. As chroot requirements change you don't have to update job specs or update artifacts on nomad clients

Disadvantages
1. Maintain 1 packer/terraform image for the host file system and N rootfs' where N=total unique nomad tasks.
2. No ecosystem over managing rootfs like versioning and hosting the rootfs.

We have gotten pretty comfortable with packer/terraform and the ecosystem around those. So as long as there is equivalent considerations to rootfs then I think it could be ok, again would just have to see the execution.

I would maybe like to see something like:
```
packer build my_job_config.rootfs // would contain same data as chroot_env from this PR
packer upload my_job.tar.gz // upload to remote store where nomad would go get and unpack into chroot
```

I'm not sure exactly what the versioning would look like, maybe as simple as prefix or postfixing file name.

**Pull request closed without merging**